src/Pulsar.Client/Api/DeadLetters.fs

```diff
@@ -0,0 +1,63 @@
+﻿namespace Pulsar.Client.Api
+
+open System
+open Pulsar.Client.Api
+open Pulsar.Client.Common
+open System.Collections.Generic
+open System.Runtime.InteropServices
+open FSharp.Control.Tasks.V2.ContextInsensitive
+open FSharp.UMX
+
+type DeadLettersPolicy(maxRedeliveryCount : int, [<Optional; DefaultParameterValue(null : string)>] deadLetterTopic : string) =
+    member __.MaxRedeliveryCount = maxRedeliveryCount
+    member __.DeadLetterTopic = deadLetterTopic
+
+type DeadLettersProcessor
+    (policy: DeadLettersPolicy,
+     getTopicName: unit -> string,
+     getSubscriptionNameName: unit -> string,
+     createProducer: string -> IProducer) =
+
+    let store = Dictionary<MessageId, ResizeArray<Message>>()
+
+    let producer = lazy (
+
+        let topicName =
+            if String.IsNullOrEmpty(policy.DeadLetterTopic) |> not then
+                policy.DeadLetterTopic
+            else
+                (sprintf "%s-%s-DLQ" (getTopicName()) (getSubscriptionNameName()))
+
+        createProducer topicName
+    )
+
+    let sendAndForgetAsync (builder : MessageBuilder) =
+        task {
+            let! _ = producer.Value.SendAsync(builder)
+            return ()
+        }
+
+    with
+        member __.ClearMessages() = store.Clear()
+
+        member __.AddMessage (message : Message) =
+            let id = message.MessageId
+            if message.RedeliveryCount |> int >= policy.MaxRedeliveryCount then
+                if store.ContainsKey(id) |> not then store.[id] <- ResizeArray<Message>()
+                store.[id].Add(message)
+
+        member __.RemoveMessage messageId =
+            store.Remove(messageId) |> ignore
+
+        member this.ProcessMessages messageId =
+            if  store.ContainsKey messageId then
+
+                store.[messageId]
+                |> List.ofSeq
```

Comment:
Why is this cast needed?

Reply:
It's unnecessary cast. Removed.

Reply:
Please inspect `ProducerImpl.SendAndForgetAsync`. Body of this method contains similar code, but method named  `SendAndForgetAsync`.

```
member this.SendAndForgetAsync (message: byte[]) =
            task {
                connectionHandler.CheckIfActive()
                let! _ = this.SendMessage (MessageBuilder(message))
                return ()
            }
```